src/core/classes/language.ts

```diff
@@ -0,0 +1,96 @@
+import { extend, cloneGrammar, resolveGrammar } from '../../shared/language-util';
+import type LanguageRegistry from './language-registry';
+import type { ComponentProtoBase } from './language-registry';
+import type { Grammar, GrammarOptions } from '../../grammar';
+
+export default class Language {
+	def: LanguageProto;
+	registry: LanguageRegistry;
+	evaluatedGrammar?: Grammar;
+
+	constructor (def: LanguageProto, registry: LanguageRegistry) {
+		this.def = def;
+		this.registry = registry;
+	}
+
+	get id () {
+		return this.def.id;
+	}
+
+	get alias () {
+		if (!this.def.alias) {
+			return [];
+		}
+
+		return Array.isArray(this.def.alias) ? this.def.alias : [this.def.alias];
+	}
+
+	get base (): Language | null {
+		if (!this.def.base) {
+			return null;
+		}
+
+		return this.registry.get(this.def.base.id) ?? null;
+	}
+
+	get extends () {
+		if (!this.def.extends) {
+			return [];
+		}
+		return Array.isArray(this.def.extends) ? this.def.extends : [this.def.extends];
+	}
+
+	get grammar (): Grammar {
+		// Lazily evaluate grammar
+		const def = this.def;
+
+		let { id, grammar } = def;
+		let base = this.base;
+
+		if (typeof grammar === 'function') {
+			grammar = grammar.call(this, {
+				base,
```

Comment:
We should probably pass the required languages as well, so we can destructure them in the `grammar()` function and use the way we can use `base`. It can be helpful, for example, when defining the `JSDoc` language: there will be no need to get the needed languages using `getLanguage()`. See https://github.com/PrismJS/prism/blob/44429cac5aa9e2ae56d8adab20195edb1496fd30/src/languages/jsdoc.ts#L11-12.